pkg/backends/calico/routes.go

```diff
@@ -16,29 +16,291 @@ package calico
+
+	"github.com/kelseyhightower/confd/pkg/resource/template"
 )
 
+const (
+	envStaticRoutes = "CALICO_STATIC_ROUTES"
+	staticRoutesKey = "calico/static-routes"
+	clusterIPStr    = "k8s-cluster-ips"
+)
+
+// routeGenerator defines the data fields
+// necessary for monitoring the services/endpoints resources for
+// valid service ips to advertise
 type routeGenerator struct {
-	client *client
+	sync.Mutex
+	client                  *client
+	nodeName                string
+	svcInformer, epInformer cache.Controller
+	svcIndexer, epIndexer   cache.Indexer
+	svcRouteMap             map[string][]string // maps service name to ip
 }
 
+// NewRouteGenerator initializes a kube-api client and the informers
 func NewRouteGenerator(c *client) (rg *routeGenerator, err error) {
-	rg = &routeGenerator{client: c}
+	// initialize empty route generator
+	rg = &routeGenerator{
+		client:      c,
+		nodeName:    template.NodeName,
+		svcRouteMap: make(map[string][]string),
+	}
+
+	// set up k8s client
+	// attempt 1: KUBECONFIG env var
+	cfgFile := os.Getenv("KUBECONFIG")
+	cfg, err := clientcmd.BuildConfigFromFlags("", cfgFile)
+	if err != nil {
+		// attempt 2: in cluster config
+		if cfg, err = rest.InClusterConfig(); err != nil {
+			return nil, err
+		}
+	}
+	client, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// set up services informers
+	svcWatcher := cache.NewListWatchFromClient(client.Core().RESTClient(), "services", "", fields.Everything())
+	svcHandler := cache.ResourceEventHandlerFuncs{AddFunc: rg.onSvcAdd, UpdateFunc: rg.onSvcUpdate, DeleteFunc: rg.onSvcDelete}
+	rg.svcIndexer, rg.svcInformer = cache.NewIndexerInformer(svcWatcher, &v1.Service{}, 0, svcHandler, cache.Indexers{})
+
+	// set up endpoints informers
+	epWatcher := cache.NewListWatchFromClient(client.Core().RESTClient(), "endpoints", "", fields.Everything())
+	epHandler := cache.ResourceEventHandlerFuncs{AddFunc: rg.onEPAdd, UpdateFunc: rg.onEPUpdate, DeleteFunc: rg.onEPDelete}
+	rg.epIndexer, rg.epInformer = cache.NewIndexerInformer(epWatcher, &v1.Endpoints{}, 0, epHandler, cache.Indexers{})
+
 	return
 }
 
+// Start reads CALICO_STATIC_ROUTES for k8s-cluster-ips
+// and CIDRs to advertise, comma-separated
 func (rg *routeGenerator) Start() (err error) {
-	// MVP implementation: read CIDRs to advertise,
-	// comma-separated, from an environment variable
-	// CALICO_STATIC_ROUTES.
-	routeString := os.Getenv("CALICO_STATIC_ROUTES")
-	cidrs := []string{}
+	var (
+		isDynamic   bool
+		ch          = make(chan struct{})
+		cidrs       = []string{}
+		routeString = os.Getenv(envStaticRoutes)
+	)
+
+	// parse routeString
 	for _, route := range strings.Split(routeString, ",") {
 		cidr := strings.TrimSpace(route)
-		if cidr != "" {
+		if len(cidr) == 0 {
+			// skip empty string
+			continue
+		} else if cidr == clusterIPStr {
+			// handle k8s-cluster-ips
+			isDynamic = true
+		} else {
+			// consider anything else as cidr
```

Comment:
We should validate the input - otherwise we might crash bird by passing it an invalid config.

Reply:
should we also validate the clusterIP or are we assured that the format of that field will be correct?

Reply:
Nah, that will be validated by k8s